runtime/globroots.c

```diff
@@ -198,70 +208,85 @@ CAMLexport void caml_remove_global_root(value *r)
   caml_delete_global_root(&caml_global_roots, r);
 }
 
+enum gc_root_class {
+  YOUNG,
+  OLD,
+  UNTRACKED
+};
+
+static enum gc_root_class classify_gc_root(value v)
+{
+  if(!Is_block(v)) return UNTRACKED;
+  if(Is_young(v)) return YOUNG;
+  if(Is_in_heap(v)) return OLD;
+  return UNTRACKED;
+}
+
 /* Register a global C root of the generational kind */
 
 CAMLexport void caml_register_generational_global_root(value *r)
 {
-  value v = *r;
   CAMLassert (((intnat) r & 3) == 0);  /* compact.c demands this (for now) */
-  if (Is_block(v)) {
-    if (Is_young(v))
+
+  switch(classify_gc_root(*r)) {
+    case YOUNG:
       caml_insert_global_root(&caml_global_roots_young, r);
-    else if (Is_in_heap(v))
+      break;
+    case OLD:
       caml_insert_global_root(&caml_global_roots_old, r);
+      break;
+    case UNTRACKED: break;
   }
 }
 
 /* Un-register a global C root of the generational kind */
 
 CAMLexport void caml_remove_generational_global_root(value *r)
 {
-  value v = *r;
-  if (Is_block(v)) {
-    if (Is_in_heap_or_young(v))
-      caml_delete_global_root(&caml_global_roots_young, r);
-    if (Is_in_heap(v))
+  switch(classify_gc_root(*r)) {
+    case OLD:
       caml_delete_global_root(&caml_global_roots_old, r);
+      /* Fallthrough: the root can be in the young list while actually
+         being in the major heap. */
```

Comment:
This fallthrough makes semantic sense because `OLD` can be understood as "OLD or maybe still YOUNG". (Nothing to change here.)